### PR TITLE
[Enhancement] support coalesce on fe functions

### PR DESCRIPTION
**Pull request by Seaven**
Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: fe/fe-core/src/test/resources/sql/query_dump/coalesce_join.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
...s_tickets`\nLEFT JOIN `basic_ticket_dimensions_2` AS `basic_ticket_dimensions`\nON `cs_tickets`.`ticket` <> '0' AND `cs_tickets`.`ticket` = `basic_ticket_dimensions`.`ticket` AND NOT `cs_tickets`.`ticket` IS NULL\nLEFT JOIN `ticket_team_dimensions_2` AS `ticket_team_dimensions`\nON `cs_tickets`.`ticket` <> '0' AND `cs_tickets`.`ticket` = `ticket_team_dimensions`.`ticket` AND NOT `cs_tickets`.`ticket` IS NULL\nLEFT JOIN `user_location_geo_2` AS `user_location_geo`\nON `cs_tickets`.`ts` <= `user_location_geo`.`ts_end` AND `cs_tickets`.`ts` >= `user_location_geo`.`ts_start` AND `cs_tickets`.`user` <> '0' AND `cs_tickets`.`user` = `user_location_geo`.`user` AND NOT `cs_tickets`.`user` IS NULL\nWHERE `cs_tickets`.`ds` <= '2023-09-08'\nAND CAST(`cs_tickets`.`ds` AS DATE) < STR_TO_DATE('2023-08-14T00:00:00', '%Y-%m-%dT%T')\nAND CAST(`cs_tickets`.`ds` AS DATE) >= STR_TO_DATE('2022-06-20T00:00:00', '%Y-%m-%dT%T')\nAND COALESCE(`user_location_geo`.`dim_origin_country`, '-unknown-') IN ('FR')\nGROUP BY  CAST(DATE_SUB(DATE_TRUNC('week',DATE_ADD(CAST(`cs_tickets`.`ds` AS DATETIME),INTERVAL 1 DAY)),INTERVAL 1 DAY) AS DATETIME)\nORDER BY `cs_payments_refund_inbound_tickets` DESC\nLIMIT 50000;\n",
```
Comment (fzhedu) [inline fe/fe-core/src/test/resources/sql/query_dump/coalesce_join.json:2]: unrelated?

**Review by packy92 — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 6 / 7 (85.71%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorFunctions.java|4|5|80.00%|[1047]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/ScalarOperatorEvaluator.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by packy92 — APPROVED** [review]
